Add 'irremote' package as a basic infra-red receiver driver (#384)

neildavis opened the pull request:
Supports only NEC (extended address) protocol including repeat codes so far

(Partially) fixes #384

neildavis committed:
Add 'irremote' package as a basic infra-red driver

Currently contains an IR Receiver device only.
An IR Sender device will be added shortly

Supports receiving only NEC (extended address) protocol including repeat codes so far

sago35 commented:
Why isn't CircleCI working?

neildavis commented:
> Why isn't CircleCI working?

I don't know. I was hoping someone here could help with that? I also asked on Slack in `#tinygo-dev` but did not get an answer there.

Is there something I need to do to get CCI to run? I set it up on my fork but that did not seem to make any difference. I assumed maybe it only runs for `tinygo-org` members?

If there is action required by contributors to make this work it would be helpful to add that to the `CONTRIBUTING` doc

deadprogram commented:
I do not know why it is not running. Some new feature of Github not quite working, perhaps?

neildavis commented:
> I do not know why it is not running. Some new feature of Github not quite working, perhaps?

@deadprogram Could you try toggling the setting in CircleCi please?
[Project Settings -> Advanced -> Build forked pull requests](https://app.circleci.com/settings/project/github/tinygo-org/drivers/advanced)
(not sure if the link will work, I can't see it due to permissions but crafted it from my fork & replacing the org name)
Assuming it's 'On' already, maybe toggling it off/on will help?
Thanks

deadprogram commented:
It was working previously. Not sure why it stopped, probably related to the GH protection settings that are now automatic and required. CircleCI does not have that setting. So, sounds like a support request to either GH or CircleCI or both.

deadprogram commented on irremote/irremote.go:
```diff
@@ -0,0 +1,173 @@
+type IRCallback func(code uint32, addr uint16, cmd uint8, repeat bool)
+
+// Public API for an IR receiver device
+type IRReceiverDevice interface {
```
Is there another device that implements this interface? Just wondering if it is better to wait until there is another before using interface here?

neildavis replied:
Yes, you're probably right. My thinking was that extending to cover other protocols (e.g. Sony SIRC, Phillips RC5/RC6 etc.) may result in additional devices in the future, but that's probably premature for now, especially on a pre-1.0 release. 

On a similar note I considered renaming the `New()` factory/constructor to `NewNEC()` for the same reason, thinking maybe it should be explicit and also reserving the generic `New()` for a potential 'automatic' protocol sensing device in the future? Again, I guess that's premature, although it does raise the question of whether we should strive to maintain backwards source compatibility in iterative versions prior to a 1.0 release? My opinion would be 'no' but happy to go with the consensus here. Any guidance?

deadprogram replied:
> whether we should strive to maintain backwards source compatibility in iterative versions prior to a 1.0 release

yes, unless we have a good reason to change. In which case better to change everything in a single release if possible.

deadprogram commented on irremote/irremote.go:
```diff
@@ -0,0 +1,173 @@
+	// Configure pins for the IR receiver module
+	Configure()
+	// Start/Stop receiving IR callbacks (pass nil to stop)
+	Callback(cb IRCallback)
```
Perhaps better to name this function something like `HandleDataComplete()`?

neildavis replied:
I agree that this is badly named and needs changing, although I'm not too keen on `HandleDataComplete()` (sorry!)  Personally I'd prefer to avoid the terminology of 'complete' in this case. My justification for this is that multiple IR commands/codes occur in discrete packets over time with no real concept of 'completing' or 'terminating' until you set the callback to `nil` to tell the driver you're not interested in them anymore. i.e. it's not a continous stream that has a natural completion. Neither is it a 'one-shot' async call that 'completes' Not sure I'm describring that too well but hopefully you can understand my thoughts?

It's a 'setter' function - you are setting a callback function property, passing `nil` to clear it. It's analagous to `machine.SetInterrupt()` and actually called from within an interrupt handler - so IMHO it should be named something of the form `SetXxxYyy()` How about  `SetCommandHandler()`, `SetCommandCallback()`,  or just plain `SetCallback()`??

On a related note I was questioning the use of using callback functions from an interrupt handler. I'm not sure if this is a good idea? However, the only alternative I can think of is to use a channel to 'transpose' the callback back to the client thread/goroutine. But use of channels would enforce the use of a scheduler which isn't great for the more constrained devices like AVR, so probably better to leave that up to the client? Any insight/advice here would be gratefully received! Thanks!

deadprogram replied:
This is usually beyond the scope of what most drivers should do. However interrupt handlers can and should use channels. "It depends" is always a sure answer.

Anyhow the word "Handler" is more commonly used in Go such as `net/http`. I like `SetCommandHandler()` the most here. :smile_cat:

neildavis commented:
@deadprogram I refactored a little in light of your comments to remove the `interface` and also tidy up the API to closer reflect the commonly used [Arduino library](https://github.com/Arduino-IRremote/Arduino-IRremote) - mainly using a `struct` for the callback data rather than a long parameter list. 

I think this is ready for [re]review now, so please take a look when you get a chance. Thanks!

neildavis commented:
@deadprogram I refactored a little in light of your comments to remove the `interface` and also tidy up the API to closer reflect the commonly used [Arduino library](https://github.com/Arduino-IRremote/Arduino-IRremote) - mainly using a `struct` for the callback data rather than a long parameter list. 

I think this is ready for [re]review now, so please take a look when you get a chance. Thanks!

neildavis commented:
@deadprogram I changed the callback terminology to `SetCommandHandler()` as discussed above. Repeat codes are also now implemented as a small addition to the protocol, and I added a line to `README.md` for the device. I'm hoping this is ready to merge now? Please take a look when you get a chance. Thanks :)

deadprogram commented:
Let's merge now so any future work can be in smaller more atomic increments.

Thank you very much for working on this @neildavis
